wgpu-core/src/device/life.rs

```diff
@@ -1,202 +1,148 @@
-        self.pipeline_layouts
-            .extend_from_slice(&other.pipeline_layouts);
-        self.render_bundles.extend_from_slice(&other.render_bundles);
-        self.query_sets.extend_from_slice(&other.query_sets);
-    }
-
-    pub(super) fn add_render_bundle_scope<A: HalApi>(&mut self, trackers: &RenderBundleScope<A>) {
-        self.buffers.extend(trackers.buffers.used());
-        self.textures.extend(trackers.textures.used());
-        self.bind_groups.extend(trackers.bind_groups.used());
-        self.render_pipelines
-            .extend(trackers.render_pipelines.used());
-        self.query_sets.extend(trackers.query_sets.used());
-    }
-
-    pub(super) fn add_bind_group_states<A: HalApi>(&mut self, trackers: &BindGroupStates<A>) {
-        self.buffers.extend(trackers.buffers.used());
-        self.textures.extend(trackers.textures.used());
-        self.texture_views.extend(trackers.views.used());
-        self.samplers.extend(trackers.samplers.used());
+impl<Id, R> ResourceMap for HashMap<Id, Arc<R>>
+where
+    Id: id::TypedId,
+    R: Resource<Id>,
+{
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+    fn as_any_mut(&mut self) -> &mut dyn Any {
+        self
+    }
+    fn clear_map(&mut self) {
+        self.clear()
+    }
+    fn extend_map(&mut self, r: &mut ResourceMaps) {
+        if let Some(other) = r.maps.get_mut(R::TYPE) {
+            if let Some(other) = other.as_any_mut().downcast_mut::<Self>() {
+                self.extend(other.drain());
+            }
+        }
     }
 }
 
-/// Raw backend resources that should be freed shortly.
-#[derive(Debug)]
-struct NonReferencedResources<A: hal::Api> {
-    buffers: Vec<A::Buffer>,
-    textures: Vec<A::Texture>,
-    texture_views: Vec<A::TextureView>,
-    samplers: Vec<A::Sampler>,
-    bind_groups: Vec<A::BindGroup>,
-    compute_pipes: Vec<A::ComputePipeline>,
-    render_pipes: Vec<A::RenderPipeline>,
-    bind_group_layouts: Vec<A::BindGroupLayout>,
-    pipeline_layouts: Vec<A::PipelineLayout>,
-    query_sets: Vec<A::QuerySet>,
+/// A struct that keeps lists of resources that are no longer needed by the user.
+#[derive(Default)]
+pub(crate) struct ResourceMaps {
+    pub(crate) maps: FastHashMap<ResourceType, Box<dyn ResourceMap>>,
 }
 
-impl<A: hal::Api> NonReferencedResources<A> {
-    fn new() -> Self {
-        Self {
-            buffers: Vec::new(),
-            textures: Vec::new(),
-            texture_views: Vec::new(),
-            samplers: Vec::new(),
-            bind_groups: Vec::new(),
-            compute_pipes: Vec::new(),
-            render_pipes: Vec::new(),
-            bind_group_layouts: Vec::new(),
-            pipeline_layouts: Vec::new(),
-            query_sets: Vec::new(),
-        }
+impl ResourceMaps {
+    fn add_type<Id, R>(&mut self) -> &mut Self
+    where
+        Id: id::TypedId,
+        R: Resource<Id>,
+    {
+        let map = HashMap::<Id, Arc<R>>::default();
```

Comment:
I haven't tested it, but wouldn't it be worthwhile here to use a faster hasher for these hashmaps, since std's siphash is quite slow?